### PR TITLE
merge date 10.06.2015 re-apply: cleanup TaskConfiguration API

### PR DESCRIPTION
**Pull request by goldhoorn**
@doudou please do a cleanup for ruby1.9.1 for this PR. I will comment the lines in follow...

### CODE REVIEW

File: lib/orocos/configurations.rb [hunk 1]
```diff
@@ -152,12 +169,13 @@ def load_from_yaml(file)
...
-
+                result = normalize_conf(YAML.load(StringIO.new(doc)) || Hash.new)
                 conf_options = options[idx].first
-                name = conf_options.delete('name')
+                name  = conf_options.delete(:name)
+                chain = conf(conf_options.delete(:chain), true)
+                result = Orocos::TaskConfigurations.merge_conf(result, chain, true)
                 changed = in_context("while loading section #{name} of #{file}") do
-                    add(name, result || Hash.new, conf_options)
+                    add(name, result, **conf_options)
```
Comment (goldhoorn) [inline lib/orocos/configurations.rb]: ** is unknown for ruby 1.9.1

File: lib/orocos/configurations.rb [hunk 2]
```diff
@@ -272,6 +292,25 @@ def add(name, conf, options = Hash.new)
...
+        #   objects which respond to #raw_read, this method returning a Typelib
+        #   value.
+        # @param [String] section_name the section name. If one already exists
+        #   with that name it is overriden
+        # @param [Boolean] merge if true, the configuration will be merged with
+        #   an existing section that has the same name (if there is one)
+        # @return [Boolean] true if the configuration changed, and false
+        #   otherwise
+        # @see add
+        def extract(name, task, merge: true)
```
Comment (goldhoorn) [inline lib/orocos/configurations.rb]: merge: true syntax invalid

File: lib/orocos/configurations.rb [hunk 3]
```diff
@@ -489,21 +586,62 @@ def conf(names, override = false)
...
+        #
+        # The objects are equivalent to the ruby objects one would get by
+        # enumerating a task's property
+        #
+        # @param [String,Array<String>] names the configurations to apply. See
+        #   {#conf} for more details
+        # @param [Boolean] override see {#conf}
+        #
+        # @see conf conf_to_typelib
+        def conf_as_ruby(names, override: false)
```
Comment (goldhoorn) [inline lib/orocos/configurations.rb]: same here : operator invalid

File: lib/orocos/configurations.rb [hunk 4]
```diff
@@ -489,21 +586,62 @@ def conf(names, override = false)
...
+        #
+        # The typelib values are equivalent to the typelib objects one would get
+        # by enumerating a task's property
+        #
+        # @param [String,Array<String>] names the configurations to apply. See
+        #   {#conf} for more details
+        # @param [Boolean] override see {#conf}
+        #
+        # @see conf conf_to_ruby
+        def conf_as_typelib(names, override: false)
```
Comment (goldhoorn) [inline lib/orocos/configurations.rb]: ```:``` again

File: lib/orocos/configurations.rb [hunk 5]
```diff
@@ -250,21 +268,23 @@ def evaluate_numeric_field(field, field_type)
...
-            conf = config_from_hash(conf)
+        # @param [{String=>Object}] conf the configuration data, as either a
+        #   mapping from property names to property values, or property names to
+        #   plain Ruby objects. It gets passed to {#normalize_conf}.
+        # @param [Boolean] merge if true, the configuration will be merged with
+        #   an existing section that has the same name (if there is one)
+        # @return [Boolean] true if the configuration changed, and false
+        #   otherwise
+        # @see extract
+        def add(name, conf, merge: true)
```
Comment (goldhoorn) [inline lib/orocos/configurations.rb]: ```:``` again

File: lib/orocos/configurations.rb [hunk 6]
```diff
@@ -646,49 +786,85 @@ def current_context
...
+        # @overload save(task, file, name)
+        #   @param [TaskContext] task the task whose configuration is to be saved
+        #   @param [String] file either a file or a directory. If it is a
+        #     directory, the generated file will be named based on the task's
+        #     model name
+        #   @param [String,nil] name the name of the new section. If nil is given,
+        #     defaults to task.name 
+        #   @return [Hash] the task configuration in YAML representation, as
+        #     returned by {.config_as_hash}
+        def self.save(config, file, name, task_model: nil)
```
Comment (goldhoorn) [inline lib/orocos/configurations.rb]: ```:``` again

**Comment by doudou** [comment]
1.9.1 compatibility in not required for the current master. Been discussed already.

**Comment by goldhoorn** [comment]
Did we? Nerarly all users i know still use ruby1.9.1 which is the default version of 14.04. I would like to prevent a brick for all this installations as long there is no real need for. We have also no check if we decide to merge this (which means kill 1.9.1) then we should first add a check to a other place (e.g. rock core package set?)

**Comment by doudou** [comment]
ruby2.0 is available on 14.04 and [yes](https://www.mail-archive.com/rock-dev@dfki.de/msg01561.html).

I agree that we have to phase it in nicely, but, you know, it is master.

**Comment by doudou** [comment]
In any case, I'm preparing a patch to rock.core/package_set to force usage of 2.0. Should we also make people use 2.0 by default in bootstrap.sh ? AFAIK it got a lot of testing.

**Comment by goldhoorn** [comment]
I would do the check in the init.rb of thr package set only if master flavor is selected. 

I would then announce the break together with the streamline header patch (and merge mid of next week)

**Comment by goldhoorn** [comment]
The bootstrap.sh is not sufficient because it can be used still for stable.

**Comment by goldhoorn** [comment]
Sorry missunderstood you. Of course bootstrap.sh should use it

**Comment by doudou** [comment]
> Sorry missunderstood you. Of course bootstrap.sh should use it

Not sure ... Modifying bootstrap.sh *will* make people using stable/rock-15.05 use 2.0 by default.

**Comment by goldhoorn** [comment]
The release has its own bootstrap.sh so no influence. For stable we should not force it (yet) the check should only force ruby larger equals 2 is master flavor is selected

**Comment by goldhoorn** [comment]
And for the bootrap script it is fine for stable to change to ruby2.0 as default

**Comment by doudou** [comment]
http://rock-robotics.org/stable/documentation/installation.html points to bootstrap.sh from master.

**Comment by goldhoorn** [comment]
We should extend the documentation with a hint to the latest release. Not sure how we should do this becaue the latest release is moving and we have no tag/id for the laest release. I would not put a fixed URL for the release into the documentation (we will forgot to update this specially on master during the release window)

**Comment by goldhoorn** [comment]
It should currently point to:
https://github.com/rock-core/buildconf/blob/rock-15.05/bootstrap.sh

Maybe we should create a redirect from the homepage to the latest one, by doing this we can update it after releasing the release. somehting like rock-robotics.org/bootstrap.sh (for master) and rock-roboics.org/stable/bootstrap.sh for the release?!

**Comment by doudou** [comment]
That sounds good.

**Comment by goldhoorn** [comment]
See: https://github.com/rock-core/base-build_server_scripts/pull/3

However this should be applied on 10.06.2015 as announced

**Comment by doudou** [comment]
I just force-pushed a few changes to (1) keep history (reverting the revert of a merge loses the history :() and (2) make sure the test suite passed (it was not ...)

**Pull request merged**